### PR TITLE
Tk/parameterized tendencies

### PR DESCRIPTION
**Pull request by tomkimpson**
Structure for the parametrized physics tendencies (draft PR)

Overkill at this stage to implement ALL tendencies (deep convection, large scale condensation, short and long wave radiation, surface fluxes, PBL etc. ). Which of these are sufficient to just get a minimal working example running?

### CODE REVIEW

File: src/humidity.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+"""
+struct compute_relative_and_saturation_humidity{NF<:AbstractFloat}     
+
+#Set of constants used in the saturation humidity calculations
+#Where do these come from? What do they represent?
+
+e0 ::Float64= 6.108e-3
+c1 ::Float64= 17.269
+c2 ::Float64= 21.875
+t0 ::Float64= 273.16 #Paxton/Chantry have a zero_c term here?
```
Comment (milankl) [inline src/humidity.jl:13]: They might have done so to avoid the Kelvin vs Celsius offset for precision reasons?

File: src/humidity.jl [hunk 2]
```diff
@@ -0,0 +1,68 @@
+
+"""
+Humidity struct containing relevant parameters and arrays
+"""
+struct compute_relative_and_saturation_humidity{NF<:AbstractFloat}
```
Comment (milankl) [inline src/humidity.jl:5]: If it's just a struct with constants why not defining them in `Constants`?

Comment (milankl) [reply src/humidity.jl:5]: Alternatively, we could collect them in a struct `HumidityConstants` (note that types, incl custom, i.e. primitive types or structs use camel font in Julia style) and use that in `Constants`. Hence we'd define and access them via

```julia
struct HumidityConstants
    e0::NF
    c1::NF
end

struct Constants
     humidity_constants = HumidityConstants(...)
end

function abc(C::Constants)
    @unpack e0 = C.humidity_constants
    ...
end
```

@unpack e0 = C.HumidityConstants

File: src/humidity.jl [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+t2 ::Float64= 7.66 
+
+end
+
+
+
+"""
+Compute the saturation specific humidity and relative hum. from specific hum. 
+"""
+function compute_relative_and_saturation_humidity!(Diag::PrognosticVariables{NF}
```
Comment (milankl) [inline src/humidity.jl:24]: As long as relative and saturation humidity isn't used elsewhere I find it redudant to add `compute_` or `calculate_`, but that might be my style?

File: src/humidity.jl [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+@unpack temp_grid,humid_grid_saturation,humid_grid_relative = Diag.gridvars
+@unpack σ_levels_full = G.geometry
+
+nlat,nlon,nlev = size(humid_grid_saturation)
+
+
+σ_levels_full
+
+#Iterate over every point. Can we vectorize this?
```
[the remark this answers is not in the document]
Comment (milankl) [reply src/humidity.jl:38]: Maybe more efficient to rewrite with
```julia
c,t = temp_grid[k,l,m] >= zero(NF) ? c1,t1 : c2,t2
humid_grid_saturation[k,l,m] = e0*exp(c * (temp_grid[k,l,m] - t0)/(temp_grid[k,l,m] - t))
```
As the `condition ? :` notation is usually easier for the compiler to optimize due to the type-stability that comes with it and the avoidance to branch off entire code blocks.

File: src/humidity.jl [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+
+        end
+    end
+end
+
+
+for m in 1:nlev
+    if σ_levels_full[m] <= 0
+        humid_grid_saturation[:,:,m]  = 622.0_*humid_grid_saturation[:,:,m] / (pres_surf_grid_exp[1,1,m] - 0.378*humid_grid_saturation[:,:,m])
```
Comment (milankl) [inline src/humidity.jl:59]: `622.0_` is invalid notation in Julia, move to `Constants` to guarantee type-stability?

File: src/parametrization_tendencies.jl [hunk 6]
```diff
@@ -1,13 +1,83 @@
-struct parametrization_tendencies{T<:AbstractFloat}
+struct parametrization_tendencies{NF<:AbstractFloat
```
Comment (milankl) [inline src/parametrization_tendencies.jl:1]: Structs are usually named with camel font

File: src/parametrization_tendencies.jl [hunk 7]
```diff
@@ -1,13 +1,83 @@
...
+
+
+    # 1. Compute the thermodynamic variables
+
+    # 1.1 Declare some surface pressure quantities
+    pres_surf_grid_exp = exp(pres_surf_grid)
+    pres_surf_grid_exp_inv = 1/pres_surf_grid_exp
+
+    #1.2 Remove negative humidity grid values
+    #TK: don't really want to iterate in this way. Is there a nice inbuilt function to do the same thing?
```
Comment (milankl) [inline src/parametrization_tendencies.jl:33]: I'll add the following to `src/utils.jl`, so you can call `clip_negatives(humid_grid)` here. My Julia gut feeling tells me this is as fast as it gets

```julia
function clip_negatives(A::AbstractArray{T}) where T
    @inbounds for i in eachindex(A)
        A[i] = max(A[i],zero(T))
    end
end
```

Comment (milankl) [reply src/parametrization_tendencies.jl:33]: With #34 you can just write `cip_negatives!(humid_grid)` 😄

File: src/humidity.jl (hunk 3 above)
Comment (tomkimpson) [reply src/humidity.jl:24]: Agreed

File: src/parametrization_tendencies.jl (hunk 7 above)
Comment (tomkimpson) [reply src/parametrization_tendencies.jl:33]: Thats great!

**Comment by white-alistair** [comment]
@tomkimpson Hi, after talking to @milankl I was thinking of doing some work on the physical parametrisations for SpeedyWeather.jl. Would you be happy for me to take over here, or would you like to collaborate somehow? Let me know what might work best for you!

**Comment by white-alistair** [comment]
@tomkimpson thanks a lot! Okay perfect, will just create a fresh branch then.

**Comment by tomkimpson** [comment]
@white-alistair A more up to date branch for the tendencies stuff is at https://github.com/milankl/SpeedyWeather.jl/pull/39, just for your information. There you can see `src/parameterized_tendencies` is empty, apart from the function definition. To match with the rest of the code, it would be good if you could leave the function arguments unchanged e.g.

```julia
function parametrization_tendencies!(Prog::PrognosticVariables{NF}, # Prognostic variables
                                      Diag::DiagnosticVariables{NF}, # Diagnostic variables
                                     M
                                     ) where {NF<:AbstractFloat}
         
...
```

**Comment by tomkimpson** [comment]
Closing - superseded by https://github.com/milankl/SpeedyWeather.jl/pull/39

**Pull request closed without merging**